refactor: Rewrite Entity Component System

aronwk-aaron opened the pull request:
The purpose of this PR it to overhaul the Entity Component System (ECS) with knowledge gained from reverse engineering the client.
Goals are to:
- Reduced code duplication between components
- Optimize hot code paths to be more efficient
- Less error prone way to Add a component to an Entity
- Use unique_ptr for Component storage so that it is clear that when an Entity is deleted, its components are too.
- Remove an ungodly amount of crusty code
- Clearer and less error prone creation and initialization of components through a constructor, and then all initialization and loading is done through and Startup and loader methods
- Remove outdated C like component getting
- Use proper delta compression for serialization
- Add tests for serialization to ensure progression and future protection
- etc (I couldn't think of anything else, this is just a fun thing to do after doing a LOT of RE)

aronwk-aaron committed:
Break out the model component
into the model, item, and mutable model behaviors

aronwk-aaron committed:
Merge branch 'main' into item-component

aronwk-aaron committed:
fix cmake

aronwk-aaron committed:
include

aronwk-aaron committed:
Merge branch 'main' into item-component

aronwk-aaron committed:
hey it compiles

EmosewaMC committed:
Rename some variables in Component

And add more virtuals

EmosewaMC committed:
I hope this works

EmosewaMC committed:
Move to shared pointer

aronwk-aaron committed:
update component names, document heirarchy

EmosewaMC committed:
holy mother of const

Use const everywhere that makes sense
return const variables when it makes sense
const functions and variables again, where it makes sense
No raw access and modifications to protected members
Move template definitions to tcc file

idk how I feel about this one

EmosewaMC committed:
Remove shared pointer, ODR of componentType variable

EmosewaMC committed:
Replace all auto with auto*

For components

EmosewaMC committed:
Rename from GetOwningEntity to GetParentEntity

EmosewaMC committed:
Rename some variables

- Add order for loading Components
- Enforce all components have Entity* in the first argument

EmosewaMC committed:
Add blank classes for some Components

Racing Stats

Minigame Control

EmosewaMC committed:
Update CMakeLists.txt

EmosewaMC committed:
Update includes for propertyEnteranceComponent

Class must be complete to be used in a MoveAssignment

EmosewaMC committed:
Further re-implement Entity::Initialize

wheee

EmosewaMC committed:
Merge branch 'main' into components-wheeeee

aronwk-aaron committed:
Rename RebuildComponent to QuickbuildComponent

EmosewaMC committed:
Add test for component Whitelists

EmosewaMC committed:
More robust tests

EmosewaMC committed:
Use std algorithms

EmosewaMC committed:
Update to actually work with component list

EmosewaMC committed:
Collectible, Item, further re-implement initialize

EmosewaMC committed:
Destroyable

EmosewaMC committed:
Continued re-integration of Entity::Initialize

EmosewaMC committed:
Quickbuild and Destroyable reintegration

EmosewaMC committed:
doesnt have a component

EmosewaMC committed:
whitespace and comments in quickbuild

EmosewaMC committed:
Comment out done code

EmosewaMC committed:
Fully re-implemented initialize

EmosewaMC committed:
Minor formatting change

EmosewaMC committed:
Remove old Entity Initialize

EmosewaMC commented on CMakeVariables.txt:
```diff
@@ -17,7 +17,7 @@ __dynamic=1
 # __compile_backtrace__=1
 # Set to the number of jobs (make -j equivalent) to compile the mariadbconn files with.
-__maria_db_connector_compile_jobs__=1
+__maria_db_connector_compile_jobs__=
```
Note: Do not ship with this.

EmosewaMC committed:
Specify auto ptr

EmosewaMC committed:
Extra * on one auto

EmosewaMC committed:
few more pointer specifiers

EmosewaMC committed:
Use only 1 script component per Entity

Serialization won't make sense if we allow two of the same component like that

EmosewaMC committed:
Use more clear control paths

EmosewaMC committed:
Better log and comment

EmosewaMC committed:
doesnt compile

aronwk-aaron committed:
update heirarchy

EmosewaMC committed:
use get

EmosewaMC committed:
remove script stuff

aronwk-aaron committed:
fix other script calls

EmosewaMC committed:
I havent checked if this compiled

EmosewaMC committed:
switch to unique_ptrs for callback timers

EmosewaMC committed:
Update Entity.cpp

EmosewaMC committed:
Finish file

EmosewaMC committed:
Update includes

EmosewaMC committed:
Remove empty destructors

EmosewaMC committed:
Fix typo

EmosewaMC committed:
Merge branch 'components-wheeeee' of https://github.com/DarkflameUniverse/DarkflameServer into components-wheeeee

EmosewaMC committed:
Another consistency pass

- change NotifyObject to use u16 string
- move stuff to header that is inline
- use u16strings instead of converting to u16 string
- move entity to dEntity

EmosewaMC committed:
Organize Entity header

Probably the third or fourth pass of this darn header...  Just keep making it better every time
Rename some functions to make more sense to a reader
Use different method for Observing/subscribing to component events
Get rid of abomination of overloading GetParentUser

EmosewaMC committed:
Merge from upstream

EmosewaMC committed:
Update DestroyableComponent.cpp

EmosewaMC committed:
Use better API terminology for radii

- SetProximityRadius just calls AddProximityRadius so its clear what is going on.
- created struct BoxDimensions for clear reading of what the floats are

EmosewaMC committed:
More organization of header

EmosewaMC committed:
Finish header re-ordering

aronwk-aaron committed:
Vendor cleanup and start Donation Vendor impl

aronwk-aaron committed:
some tweaks

aronwk-aaron committed:
make inheritence clearer

aronwk-aaron committed:
scaffold activity based components

aronwk-aaron committed:
Merge branch 'main' into components-wheeeee

aronwk-aaron committed:
AchievementVendor scaffold

aronwk-aaron committed:
Merge branch 'item-component' into components-wheeeee

aronwk-aaron committed:
rough logic around loading physics
as well as modelbehaviors

aronwk-aaron committed:
group physicstogether in heirarchy

aronwk-aaron committed:
move vars to be local

EmosewaMC committed:
Merge upstream ghosting candidate changes

EmosewaMC committed:
Remove commented out logic, add ghosting exclusion

We dont create items that are in an inventory, so that path is pointless for us.

EmosewaMC committed:
Entity work

- Add in bool cheks
- Fix component class files so they compile and link
- Fin inheritance

EmosewaMC committed:
Use better naming

- Remove use of Base.  It is implied if you inherit that the class inherited from is a Base.
- Fix compilation errors from said change.

EmosewaMC committed:
CombatAI and Vendor

EmosewaMC committed:
Bouncer cleanup

- Correct variable names
- Make serialization more efficient and make sense
- remove unneeded members
- more descriptive logs
- Move work to Startup

EmosewaMC committed:
Use of final

EmosewaMC committed:
Buff Component fixes

aronwk-aaron committed:
rename possessor to possession

aronwk-aaron committed:
Update activing and racing components
Hopefully not going to rename them for a third time, no way

EmosewaMC committed:
BuildBorderCleanup

EmosewaMC committed:
Merge branch 'components-wheeeee' of https://github.com/DarkflameUniverse/DarkflameServer into components-wheeeee

EmosewaMC committed:
Character fixes - get it compiling again

- Pass componentID to activity component constructor
- use int componentid so -1 can denote no component

EmosewaMC committed:
Component serialization updates

- Fix serialization in multiple components so they don't get dirty flags reset when it was not intentional

EmosewaMC committed:
move ctor to header for collectible

EmosewaMC committed:
Move empty definitions to header

EmosewaMC committed:
Donation Component serialization

EmosewaMC committed:
HavokVehicleComponent improvements

serialization, moving stuff to header, removal of update as we serialize dirty position so often this check is pointless.

EmosewaMC committed:
Add default constructor for EquipmentItem

Change struct constructors to proper constructors.

aronwk-aaron commented on dGame/dComponents/HavokVehiclePhysicsComponent.h:
```diff
@@ -109,9 +106,9 @@ class HavokVehiclePhysicsComponent : public Component {
-	void SetDirtyPosition(bool val);
-	void SetDirtyVelocity(bool val);
-	void SetDirtyAngularVelocity(bool val);
+	void SetDirtyPosition(bool val) { m_DirtyPosition = val; }
```
Same here, check for a diff and mark dirty

EmosewaMC committed:
InventoryComponent pass

- reduce scope usage
- bouncing returns
- std::for_each in some contexts
- extra nullptr checks
- constiness
- extra logs
- scoped enum type for eItemType
- lol serialization.

aronwk-aaron commented on dGame/dComponents/InventoryComponent.cpp:
```diff
@@ -79,28 +79,29 @@ InventoryComponent::InventoryComponent(Entity* parent, tinyxml2::XMLDocument* do
 
-		if (!subItems.empty()) {
-			const auto subItemsSplit = GeneralUtils::SplitString(subItems, ',');
+		if (subItems.empty()) return;
```
should be continue, not return
If I'm following correctly

EmosewaMC replied:
Yes

aronwk-aaron committed:
Add component order array
fix loop in inven comp

EmosewaMC committed:
ItemComponent Pass

make use of Writing a length
add C include guard

EmosewaMC committed:
Fix compilation error; add final specifiers

EmosewaMC committed:
LevelProgressionComponent cleanup

EmosewaMC committed:
LUPExhibitComponent cleanup

aronwk-aaron committed:
Merge branch 'main' into components-wheeeee

EmosewaMC committed:
MissionComponent pass

EmosewaMC committed:
MissionOfferComponent pass

EmosewaMC committed:
ModelBehaviorComponent pass

EmosewaMC committed:
SwitchComponent pass

EmosewaMC committed:
ModuleAssemblyComponent pass

EmosewaMC committed:
Merge branch 'components-wheeeee' of https://github.com/DarkflameUniverse/DarkflameServer into components-wheeeee

EmosewaMC committed:
Merge branch 'components-wheeeee' of https://github.com/DarkflameUniverse/DarkflameServer into components-wheeeee

EmosewaMC committed:
add final

EmosewaMC committed:
MovementAiComponent pass

EmosewaMC committed:
I am throwing this one out later.

EmosewaMC committed:
MultiZoneEntranceComponent pass

EmosewaMC committed:
MutableModelBehaviorComponent pass

EmosewaMC committed:
PetComponent pass

EmosewaMC committed:
PhantomPhysicsComponent pass

EmosewaMC committed:
PlayerForcedMovementComponent pass

EmosewaMC committed:
PossessableComponent pass

EmosewaMC committed:
PhantomPhysicsComponent pass

EmosewaMC committed:
PossessionComponent pass

EmosewaMC committed:
PropertyComponent pass

EmosewaMC committed:
push no pass

EmosewaMC committed:
idek if ill keep this but yeah

EmosewaMC committed:
PropertyVendorComponent pass

EmosewaMC committed:
ProximityMonitorComponent re-pass

EmosewaMC committed:
RailActivatorComponent pass

EmosewaMC committed:
local work

aronwk-aaron commented on dGame/dComponents/PropertyEntranceComponent.h:
```diff
@@ -6,13 +6,20 @@
 #include "PropertySelectQueryProperty.h"
 
+enum class ePropertySortType : int32_t {
+	SORT_TYPE_FRIENDS = 0,
```
does this need `SORT_TYPE` in front of each value?

EmosewaMC replied:
no it does not.  That can be its own PR if you can take care of that

aronwk-aaron commented on dGame/dComponents/RailActivatorComponent.cpp:
```diff
@@ -13,12 +13,21 @@
+	m_DamageImmune = m_ParentEntity->GetVar<bool>(u"rail_activator_damage_immune");
+	m_NoAggro = m_ParentEntity->GetVar<bool>(u"rail_no_aggro");
+	m_NotifyArrived = m_ParentEntity->GetVar<bool>(u"rail_notify_activator_arrived");
+	m_ShowNameBillboard = m_ParentEntity->GetVar<bool>(u"rail_show_name_billboard");
```
I've wondered, does this need to talk to the Player Forced movement component. that has two bools that are serialized ("is on rail" and "show billboard"), which this has vars for

aronwk-aaron commented:
This PR wont be merged, but remains a draft since there's a lot of research in this PR that will be implemented in smaller chunks
